File: Dynamic Programming/boj2839.py
n = int(input())

d = [5001]*(n+1)
d[0] = 0

sugar_bags = [3, 5]

for sugar_bag in sugar_bags:
    for j in range(sugar_bag, n+1):
        if d[j-sugar_bag] != 5001:
            d[j] = d[j-sugar_bag] + 1
            
print(d[n] if d[n] != 5001 else -1)

File: Dynamic Programming/BOJ2839.java
import java.util.Scanner;

public class BOJ2839 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();

        int[] dp = new int[n+1];
        for(int i = 0; i < n+1; i++) {
            dp[i] = -1;
        }
        dp[0] = 0;

        int[] sugarBags = {3, 5};

        for(int sugarBag : sugarBags) {
            for(int i = sugarBag; i < n+1; i++) {
                if(dp[i-sugarBag] != -1)
                    dp[i] = dp[i-sugarBag] + 1;
            }
        }
        System.out.println(dp[n]);
    }
}
